app/controllers/foreman_tasks/tasks_controller.rb

```diff
@@ -72,18 +72,14 @@ def resume
     def unlock
       task = find_dynflow_task
       if task.paused?
-        task.state = :stopped
-        task.save!
-        render json: { statusText: 'OK' }
+        force_unlock(task)
       else
         render json: {}, status: :bad_request
       end
     end
 
-    def force_unlock
-      task       = find_dynflow_task
-      task.state = :stopped
-      task.save!
+    def force_unlock(task = find_dynflow_task)
+      task.halt
```

Comment:
After description in https://github.com/Dynflow/dynflow/pull/406 I understand that this means:
>  for this task do halt the execution plan but please do finish any running steps.

Is this correct? If so, I still don't quite understand how the whole task will be marked as stopped since we remove here the `task.state = :stopped`? Won't it be `pending` even after the running steps are done?

Also, we kinda change here behaviour we had previously. Wouldn't it confuse the users even more?

Reply:
> Is this correct?

In general, yes, but if I were to be pedantic, I'd have to say almost. There might be different interpretations of "finish any running steps" with slight nuances, but yes. If an action's `#run` is being executed while the execution plan is halted, the execution plan moves over to stopped, but the action's `#run` will continue to run until it returns. This doesn't necessarily mean the step itself will be finished.

> I still don't quite understand how the whole task will be marked as stopped 

There is a [mechanism](https://github.com/theforeman/foreman-tasks/blob/master/app/lib/actions/base.rb#L5), which makes tasks to be updated based on the state of its corresponding execution plan as the execution plan changes states. Ideally, tasks should be read-only and the only changes should be done to them by this mechanism. Once we flip the execution plan over to stopped, this hook will trigger and update the task to stopped.

> we kinda change here behaviour we had previously. Wouldn't it confuse the users even more?

The previous behaviour was seriously broken. Dynflow doesn't care about tasks' attributes at all, so if we went in and updated the state of a task, it would not really do anything. The task would appear as stopped for some time, but if the underlying execution plan was still running, it would overwrite task's status again. It worked for things like locks which relied on "is a resource locked with a locked by a task which is not stopped", but we reworked that eventually.